New blog post about kotlin, ktor and keycloak (#105)
lunatech-labs/lunatech-blog: posts/2023-06-07-kotlin-ktor-keycloak.adoc

:imagesdir: ../media/2023-06-07-kotlin-ktor-keycloak
:lang: en
:tags: [kotlin, ktor, keycloak]

At Lunatech we build many internal tools that facilitate internal processes like keeping track of employees' vacations, share internal events and worshops and even one just for the days Lunatech offers us lunch ;) So, it's only natural that we end up creating integrations between these applications.

The goal of this post is to explore the scenario where one application authenticates with another one by using the client credentials flow, in Kotlin, Ktor and Keycloak.
To achieve this we are going to build a simple `luna-vacation-bot` that will inform you, personally, every day via a Slack message, whether the present day is a working day or a day off. For this we will need to implement a `luna-vacation-api` as well.

.Happy scenario diagram
image::diagram.png[Diagram, 600]

The `luna-vacation-api` will only reply successfully to a request if the requester also provides a valid token.
Keycloack is the middle man that will provide the tokens to `luna-vacation-bot` and that will validate the tokens received by `luna-vacation-api`.

=== How to setup Keycloak?

> First of all we need a running instance of Keycloak. Recently Lunatech deployed its own Keycloak instance for the Identity and Access Management of applications, but for this example I'm going to run a local instance of Keycloak. One can easily be run using https://www.keycloak.org/getting-started/getting-started-docker[docker].

Reviewer: ```suggestion
First, we need a running instance of Keycloak. Recently, Lunatech deployed its own Keycloak instance for the Identity and Access Management of applications, but for this example I’m going to run a local instance of Keycloak. We can easily run one using https://www.keycloak.org/getting-started/getting-started-docker[docker].
```

--data-urlencode 'client_secret=<replace-by-secret>'
----

The token will look something like:

[source,json]
----
{"access_token":"eyJhbG...75Eg","expires_in":300,"refresh_expires_in":0,"token_type":"Bearer","not-before-policy":0,"scope":"profile email"}
----

==== Check the token in jwt.io
In https://jwt.io/[jwt.io] you can inspect the contents of a token. Copy the `access_token` string and paste it in the `Encoded` box. I'm not going into details about the format and content of the token, you can read a nice https://jwt.io/introduction[introduction] about it in jwt.io site as well.

.Some of the token payload details
image::keycloak_token.png[Client secret, 400]

We can confirm that the `issuer` of the token was `lunarealms` and that the `authorized party` was `luna-vacation-bot`

=== Create luna-vacation-api
We are using Kotlin and Ktor to build the API. Following the recent tendency Ktor provides a project scaffolding https://start.ktor.io/[generator] that adds some sample code and allow to add plugins already. We want to add the Authentication JWT plugin.

.Generate luna-vacation-api project
image::ktor-project-generator.png[Client secret, 400]

> Download and open it on you IDE of choice. The test dependencies and test folder can be removed.
> Following the documentation in https://ktor.io/docs/jwt.html[ktor.io] on how to handle JSON web tokens let's add the following dependencies in `build.gradle.kts`:

Reviewer: ```suggestion
Download and open the project in your preferred IDE. You can remove the test dependencies and test folder.
Following the documentation in https://ktor.io/docs/jwt.html[ktor.io] on how to handle JSON web tokens, let's add the following dependencies in `build.gradle.kts`:
```

[source,kotlin]
----
implementation "io.ktor:ktor-server-auth-jwt:2.3.0"
implementation "io.ktor:ktor-server-auth-jwt-jvm:2.3.0"
----

We can now start adding some code. Let's start by our API endpoint `imionvacation` that determines that even day numbers are days off and odd days are working days.

[source,kotlin]
----
fun Application.configureRouting() {
    routing {
        authenticate("auth-jwt") {
            get("/imionvacation") {
                val today = LocalDate.now()
                if (today.dayOfMonth % 2 == 0) {
                    call.respondText("Today you have a day off")
                } else {
                    call.respondText("Today you have to work")
                }
            }
        }
    }
}
----

You can notice that the endpoint already includes an authentication precondition called `auth-jwt`.

> This detail is very import, as it assures that the endpoint is indeed secured by a valid JWT token.

Reviewer: ```suggestion
This detail is very import, as it ensures that a valid JWT token indeed secured the endpoint.
```

> In `fun Application.module()` we need to setup how this authentication takes places.

Reviewer: ```suggestion
In `fun Application.module()`, we need to set up how this authentication takes place.
```

> In our case we can validate the token realm, issuer, audience and also limit the only authorized party to `luna-vacation-bot`. The token signature also needs to be validated and for that we need to provide the url to the JSON Web Keys available in the `protocol/openid-connect/certs` endpoint.

Reviewer: ```suggestion
In our case, we can validate the token realm, issuer, audience and limit access to `luna-vacation-bot` for authorised parties only. The token signature also needs to be validated and for that we need to provide the url to the JSON Web Keys available in the `protocol/openid-connect/certs` endpoint.
```

[source,kotlin]
----
install(Authentication) {
        jwt("auth-jwt") {
            realm = "lunarealm"
            verifier(
                UrlJwkProvider(URL("http://localhost:8080/realms/lunarealm/protocol/openid-connect/certs")),
                "http://localhost:8080/realms/lunarealm",
            ) {
                withAudience("account")
                withClaim("azp", "luna-vacation-bot")
            }
            challenge { _, _ ->
                call.respond(HttpStatusCode.Unauthorized, "Token is not valid or has expired")
            }
            validate { credential ->
                validateCredential(credential, issuer)
            }
        }
    }
----

> If the authentication fails it returns the `challenge`, in this case a 401 Unauthorized with an error message.

Reviewer: ```suggestion
If the authentication fails, it returns the `challenge`, in this case a _401 Unauthorized_ with an error message.
```

The `validateCredentials` function is simply validating that the token has not expired, after all other validations have passed.

[source,kotlin]
----
fun validateCredential(credential: JWTCredential, issuer: String): JWTPrincipal? {
    if (credential.expiresAt?.after(Date()) == true
    ) {
        return JWTPrincipal(credential.payload)
    }

    return null
}
----

==== Test the API
We can test the api using curl.

> If we secured the endpoint properly calling the `imionvacation` should return an error.

Reviewer: ```suggestion
If we secured the endpoint properly, calling the `imionvacation` should return an error.
```

[source,commandline]
----
$ curl localhost:4040/imionvacation
Token is not valid or has expired
----

> As the diagram at the beginning of this post explained we need to request a token to Keycloak that we send together with the request, and for that we impersonate the `luna-vacation-bot`:

Reviewer: ```suggestion
As the diagram at the beginning of this post explained, we need to request a token to Keycloak that we send together with the request. For that, we impersonate the `luna-vacation-bot`:
```

[source,commandline]
----
curl -X post 'http://localhost:8080/realms/lunarealm/protocol/openid-connect/token' \
--header 'Content-Type: application/x-www-form-urlencoded' \
--data-urlencode 'grant_type=client_credentials' \
--data-urlencode 'client_id=luna-vacation-bot' \
--data-urlencode 'client_secret=<replace-by-secret>'
----

> Grab the `access_token` part of the response and add it to the request header as a Bearer token:

Reviewer: ```suggestion
We grab the `access_token` part of the response and add it to the request header as a Bearer token:
```

[source,commandline]
----
$ curl 'localhost:4040/imionvacation' \
--header 'Authorization: Bearer eyJhbGciOiJ...QBhNiX6w'
----

And now we get a result:

[source,commandline]
----
Today you have to work
----

=== Create luna-vacation-bot

For the bot app we are going to use Intellij (or your favorite IDE) and create a new Kotlin application, with gradle and the latest JVM you have available (I'm using JDK 17). After the project is created you can remove the test folder and the test dependencies.

==== Integrate with luna-vacation-api

For the http request let's use `ktor-client` and add the necessary dependencies to `build.gradle.kts`:

[source,kotlin]
----
implementation("io.ktor:ktor-client-core:2.3.1")
implementation("io.ktor:ktor-client-cio:2.3.1")
implementation("io.ktor:ktor-client-serialization:2.3.1")
implementation("io.ktor:ktor-client-content-negotiation:2.3.1")
implementation("io.ktor:ktor-serialization-kotlinx-json:2.3.1")
----

To create an HttpClient we need to add the json `ContentNegotiation` and the flag `ignoreUnknownKeys` for the response deserialization,

> where we will ignore the json fields that we are not interested in.

Reviewer: This sentence is very long... Would this be a correct modification?
```
To create an HttpClient we need to add the json `ContentNegotiation` and the flag `ignoreUnknownKeys` for the response deserialization. This will ignore the json fields that we are not interested in.
```

[source,kotlin]
----
    val client = HttpClient(CIO) {
        install(ContentNegotiation) {
            json(
                Json {
                    ignoreUnknownKeys = true
                },
            )
        }
    }
----

First we need to jwt token from Keycloak:

[source,kotlin]
----
@Serializable
data class BotToken(val access_token: String)

val token = client.submitForm(
        url = "http://localhost:8080/realms/lunarealm/protocol/openid-connect/token",
        formParameters = parameters {
            append("grant_type", "client_credentials")
            append("client_id", "luna-vacation-bot")
            append("client_secret", "<replace by client_secret>")
        },
    ).body<BotToken>().access_token
----

and then we can finally call the `luna-vacation-api`:

[source,kotlin]
----
val response = client.get("http://localhost:5050/imionvacation") {
        bearerAuth(token)
    }.bodyAsText()
----

> And what do we do with the `response` now? We send it to a slack channel using a Slack App. So let's do that next.

Reviewer: ```suggestion
We can now send the `response` to a slack channel using a Slack App. So let's do that next.
```

=== Create a Slack app
You can create your own workspace in https://slack.com/[slack.com]. Once you have that you can create a new https://api.slack.com/apps[App]. Choose `Create New App` and then `From scratch`:

.Create a new Slack App
image::slack_create_app_1.png[Diagram, 500]

I named the app `ImIOnVacation` and added it to my personal workspace:

.Create a new Slack App
image::slack_create_app_2.png[Diagram, 500]

> To finalize the App creation you need to allow some permissions:

Reviewer: allow _changed_ to _set_ (alternative would be" _change_ some permissions")
```suggestion
To finalize the App creation you need to set some permissions:
```

.Setup Slack App permissions
image::slack_app_permissions.png[Diagram, 500]

In `Basic Information` you can personalize your app appearance, but for now let's jump to `OAuth & Permission` and from there scroll down to the `Scopes` section. In `User Token Scopes` add the scopes `chat:write`:

.Setup Slack App Scopes
image::slack_app_scopes.png[Diagram, 500]

Then scroll up and click `Install to Workspace`. You will come across a permissions screen. Click `Allow`. You can see that a `User OAuth Token` was created as well.

.Slack App User OAuth token
image::slack_user_oauth_token.png[Diagram, 500]

We will need this `User OAuth Token` to be able to send messages to Slack via this App.

=== Put it all together

We can now send that `luna-vacation-api` response to a Slack channel. The channel I'll use is the chat with myself.

> Channel ids can be seen in Slack when you select a channel or a direct message chat
> and then click on the top details of the channel/person. At the very bottom you can see "Channel ID".

Reviewer: ```suggestion
We can see Channel ids in Slack when we select a channel or a direct message chat
and then click on the top details of the channel/person. At the very bottom you can see "Channel ID".
```

> For the communication with Slack we will use Slack's https://slack.dev/java-slack-sdk/guides/getting-started-with-bolt[Bolt SDK],
> starting by adding the necessary dependencies:

Reviewer: ```suggestion
For the communication with Slack we will use Slack's https://slack.dev/java-slack-sdk/guides/getting-started-with-bolt[Bolt SDK]. Let's start by adding the necessary dependencies:
```

= JWT token validation in Ktor with Client Credentials Flow
LeonorLunatech
v1.0, 2023-06-07
:title: JWT token validation in Ktor with Client Credentials Flow
:imagesdir: ../media/2023-06-07-kotlin-ktor-keycloak
:lang: en
:tags: [kotlin, ktor, keycloak]

At Lunatech we build many internal tools that facilitate internal processes like keeping track of employees' vacations, share internal events and workshops and even one just for the days Lunatech offers us lunch ;) So, it's only natural that we end up creating integrations between these applications.

The goal of this post is to explore the scenario where one application authenticates with another one by using the client credentials flow, in https://kotlinlang.org/[Kotlin], https://ktor.io/[Ktor] and https://www.keycloak.org/[Keycloak].

> To achieve this, we are going to build a simple personal `luna-vacation-bot` that will inform you daily, via a https://slack.com/[Slack] message, whether the present day is a vacation day or not. For this we will need to implement a dummy `luna-vacation-api` as well.

Author: I know I prefixed the projects with `luna` but in reality this system is not bounded to Lunatech, so I think that adding that it will `inform a Lunatech employee`  may give the feeling that no one else can try to run these projects locally

The admin console can be found at http://localhost:8080[localhost:8080] with both username and password being `admin`.

==== Create a realm
On top of the default `master` realm let's create our own realm called `lunarealm`.

.Create a new Realm
image::keycloack_create_realm.png[Create Realm, 500]

==== Create a client
Let's create the client `luna-vacation-bot` which is the one that is going to need to authenticate with the `luna-vacation-api` and use the API:

.Create a new client
image::keycloak_new_client_1.png[Create client, 500]

After setting the client id, name and description, click Next for the capability config section and turn `Client authentication` and `Service accounts roles` on and all the other authentication flows off. The client is now configured for the Client Credentials Flow in OAuth2. Click Next again and Save.

.Configure the client capability
image::keycloak_new_client_2.png[Create client, 550]

In order to increase our client security, we are also creating a new `Client Scope` called luna-vacation-bot scope with the following name and settings:

.Create a new Client Scope
image::keycloak_create_client_scope_1.png[Create client, 300]

> Once you save the new scope, a couple of configuration tabs will show up: `Mappers` and `Scope`. Go to `Mappers` and choose `Configure a new mapper` and choose `Audience`:

Author: I use it to sign that it continues with the image illustration. I think that was none there. But now there is 😄

        jwt("auth-jwt") {
            realm = "lunarealm"
            verifier(
                UrlJwkProvider(URL("http://localhost:8080/realms/lunarealm/protocol/openid-connect/certs")),
                "http://localhost:8080/realms/lunarealm",
            ) {
                withAudience("luna-vacation-bot")
                withClaim("azp", "luna-vacation-bot")
            }
            challenge { _, _ ->
                call.respond(HttpStatusCode.Unauthorized, "Token is not valid or has expired")
            }
            validate { credential ->
                validateCredential(credential, issuer)
            }
        }
    }
----
If the authentication fails it returns the `challenge`, in this case a 401 Unauthorized with an error message.

The `validateCredentials` function is simply validating that the token has not expired, after all other validations have passed.

[source,kotlin]
----
fun validateCredential(credential: JWTCredential, issuer: String): JWTPrincipal? {
    if (credential.expiresAt?.after(Date()) == true
    ) {
        return JWTPrincipal(credential.payload)
    }

    return null
}
----

==== Test the API
We can test the api using curl.

> If we secured the endpoint properly, calling the `imionvacation` endpoint without a token should return an error.

Author: `I'm I on vacation ?` 😬

> If we secured the endpoint properly, calling the `imionvacation` endpoint without a token should return an error.

Author: I'll call it `onvacation`

[source,commandline]
----
$ curl localhost:4040/imionvacation
Token is not valid or has expired
----

> As the diagram at the beginning of this post explained, we need to request a token from Keycloak and send it together with the request, and for that we impersonate the `luna-vacation-bot`:

Author: I can just say that ` We can use the credentials we have for `luna-vacation-bot``

[source,commandline]
----
curl -X post 'http://localhost:8080/realms/lunarealm/protocol/openid-connect/token' \
--header 'Content-Type: application/x-www-form-urlencoded' \
--data-urlencode 'grant_type=client_credentials' \
--data-urlencode 'client_id=luna-vacation-bot' \
--data-urlencode 'client_secret=<replace-by-secret>'
----

Grab the `access_token` part of the response and add it to the request header as a Bearer token:

[source,commandline]
----
$ curl 'localhost:4040/imionvacation' \
--header 'Authorization: Bearer eyJhbGciOiJ...QBhNiX6w'
----

And now we get a result:

[source,commandline]
----
Today you have to work
----

=== Create the luna-vacation-bot

> To create the `luna-vacation-bot` project we are going to use https://www.jetbrains.com/idea/[Intellij] (or your favorite IDE) and create a new Kotlin application, with https://gradle.org/[Gradle] and the latest JVM you have available (I'm using JDK 17). After the project is created you can remove the test folder and the test dependencies.

Reviewer 2: ```suggestion
To create the `luna-vacation-bot` project we are going to use https://www.jetbrains.com/idea/[Intellij] (or your favorite IDE) and create a new Kotlin application, with https://gradle.org/[Gradle] and the latest JVM you have available (I'm using JDK 17). After creating the project we can remove the test folder and the test dependencies.
```
I don't think saying "we are going to use.... (or your favorite IDE)". Stick to using one pronoun and avoid switching from first-plural to second person.

Maybe instead of saying "the latest JVM you have", provide the minimum JVM version requirement.

Author: what do you think of 
```To create the `luna-vacation-bot` project we are going to use https://www.jetbrains.com/idea/[Intellij] and create a new Kotlin application, with https://gradle.org/[Gradle] and JDK 17 (you can use JDK 8 or higher).```

==== Integrate with luna-vacation-api

> In order to query the `luna-vacation-api` we need a http client. We are going to use the https://ktor.io/docs/create-client.html[ktor-client] library by adding the necessary dependencies to `build.gradle.kts`:

Author: ah... the editor says it's error. But it's what makes sense to me as well. I trust the editor too much :)

[source,kotlin]
----
implementation("io.ktor:ktor-client-core:2.3.1")
implementation("io.ktor:ktor-client-cio:2.3.1")
implementation("io.ktor:ktor-client-serialization:2.3.1")
implementation("io.ktor:ktor-client-content-negotiation:2.3.1")
implementation("io.ktor:ktor-serialization-kotlinx-json:2.3.1")
----

> Besides creating the http client we also need to add the json `ContentNegotiation` and the flag `ignoreUnknownKeys` for the response deserialization,

Author: ```When creating the http client we need to add the json `ContentNegotiation` and the flag `ignoreUnknownKeys` for the response deserialization, to ignore the json fields that we are not interested in.```?

----

Now that the http client is properly setup we will start by retrieving a JWT token from Keycloak:

[source,kotlin]
----
@Serializable
data class BotToken(val access_token: String)

val token = client.submitForm(
        url = "http://localhost:8080/realms/lunarealm/protocol/openid-connect/token",
        formParameters = parameters {
            append("grant_type", "client_credentials")
            append("client_id", "luna-vacation-bot")
            append("client_secret", "<replace by client_secret>")
        },
    ).body<BotToken>().access_token
----

and then we can finally call the `luna-vacation-api`:

[source,kotlin]
----
val response = client.get("http://localhost:5050/imionvacation") {
        bearerAuth(token)
    }.bodyAsText()
----

And what do we do with the `response` now? We send it to https://slack.com/[Slack] using a https://slack.com/apps[Slack App]. So let's do that next.

=== Create a Slack app
You can create your own workspace at https://slack.com/[slack.com]. Once you have it you can create a new https://api.slack.com/apps[App]. Choose `Create New App` and then `From scratch`:

.Create a new Slack App
image::slack_create_app_1.png[Diagram, 400]

> I named the app `ImIOnVacation` and added it to my personal workspace:

Author: the idea is that the bot the bot will answer the question "I'm I on vacation?"

.Configure a Slack App worskpace
image::slack_create_app_2.png[Diagram, 450]

To finalize the App creation you need to set some permissions:

.Setup Slack App permissions
image::slack_app_permissions.png[Diagram, 400]

> In `Basic Information` you can personalize your app appearance, but for now let's jump to `OAuth & Permission` and, from there, scroll down to the `Scopes` section. In `User Token Scopes` add the scopes `chat:write`:

Reviewer 2: ```suggestion
In `Basic Information` you can personalize the appearance of the App, but for now let's jump to `OAuth & Permission` and, from there, scroll down to the `Scopes` section. In `User Token Scopes` add the scopes `chat:write`:
```

.Setup User Token Scopes
image::slack_app_scopes.png[Diagram, 500]

> Then scroll up and click `Install to Workspace`. That will take you to a permissions screen: click `Allow`. You can see that a `User OAuth Token` was created as well.

Reviewer 2: ```suggestion
Then scroll up and click `Install to Workspace`. That will take us to a permissions screen: click `Allow`. We can see that a `User OAuth Token` was created as well.
```

`you` or `we`? For the sake of consistency, I'd say use first-person pronouns throughout the blog post.

.Slack App User OAuth token
image::slack_user_oauth_token.png[Diagram, 500]

We will need this `User OAuth Token` to be able to send messages to Slack via this App.

=== Put it all together

We can now send that `luna-vacation-api` response to a Slack channel. The channel I'll use is the chat with myself.

> In Slack, channel ids can be seen when you select a channel or a direct message chat
> and then click on the top details of the channel/person. At the very bottom you can see "Channel ID".

Reviewer 2: ```suggestion
On Slack, channel IDs can be seen when you select a channel or a direct message chat
and then click on the top details of the channel/person. At the very bottom you can see "Channel ID".
```

I think that this "when you select a channel or a direct message chat
and then click on the top details of the channel/person" is a bit confusing.

Author: I'm going to simplify it with `On Slack, channel IDs can be seen on the channel or chat details, at the very bottom.`

For the communication with Slack we will use Slack's https://slack.dev/java-slack-sdk/guides/getting-started-with-bolt[Bolt SDK],
starting by adding the necessary dependencies:

[source,kotlin]
----
implementation("com.slack.api:bolt:1.29.2")
implementation("com.slack.api:bolt-servlet:1.29.2")
implementation("com.slack.api:bolt-jetty:1.29.2")
implementation("org.slf4j:slf4j-simple:1.7.36")
----

Now let's https://slack.dev/java-slack-sdk/guides/getting-started-with-bolt[compose our message].
The function `getIfImOnVacation` is wrapping the `luna-vacation-api` request.

[source,kotlin]
----
val text = getIfImOnVacation(client)
val response = slack.methods("xoxp-...").chatPostMessage { req: ChatPostMessageRequestBuilder ->
    req
        .channel("<replace by my channel>")
        .text(text)
}
----

> Make sure `luna-vacation-api` is still running and now let's run `luna-vacation-bot`. If you check Slack you should see a message:

Reviewer 2: ```suggestion
Make sure `luna-vacation-api` is still running and now let's run `luna-vacation-bot`. If we check Slack we should see a message:
```

.Slack message
image::slack_message.png[Diagram, 300]

> The complete source code for the https://github.com/LeonorLunatech/luna-vacation-api[luna-vacation-api] and https://github.com/LeonorLunatech/luna-vacation-bot[luna-vacation-bot] are available in GitHub.

Reviewer 2: ```suggestion
The complete source code for the https://github.com/LeonorLunatech/luna-vacation-api[luna-vacation-api] and https://github.com/LeonorLunatech/luna-vacation-bot[luna-vacation-bot] are available on GitHub.
```

=== Next steps

We can transform this kind of application integration in a real automation by running it periodically.
A cron job allows to achieve this goal and the http://www.quartz-scheduler.org/documentation/quartz-2.3.0/quick-start.html[quartz-scheduler] library integrates well with Kotlin.
The cron job can be configured as often was one wishes like several time a day or once a month, for example.

> In this post we used a completely dummy API. For a real use case we would replace it by an API that provides data reports, for example.

Reviewer 2: ```suggestion
In this blog post we used a dummy API. For a real use case we could replace it with an API that provides data reports, for example.
```
Not sure the word `completely` is needed. `Dummy` is good enough I think.

Author: My mediterranean blood.. exaggerating things a bit 😄
